mautic/mautic
# TPROD-344: Update to Symfony 5

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:



This PR will contain the minimal steps to switch to Symfony 5, without updating any non-crucual dependency.
This is on purpose, to ensure we can address issues one at a time, and not have to take into account side effects from other dependencies.

* minimal changes
    * [minimal dependency changes to switch to symfony 5](https://github.com/mautic/mautic/pull/12213/commits/ad4d991ea4f042117bd2bad822c52bca8ddf1ded)
        ```
        composer update leezy/pheanstalk-bundle litesaml/lightsaml lightsaml/sp-bundle symfony/*
        ```
    * [minimal config changes to be able to bootstrap Mautic](https://github.com/mautic/mautic/pull/12213/commits/d5137701ee7d522b325bfb0eda738ddfb14b71c5)
    * [minimal changes to be able to run tests (regardless of failures)](https://github.com/mautic/mautic/pull/12213/commits/b22252e671e9d9d0f6f0d7d3c81e47772dc39d42)
* a lot of changes to fix the tests and PHPSTAN compatibility (most important ones listed below)
    * [ensure that EventDispatcherInterface::dispatch() calls return the event](https://github.com/mautic/mautic/pull/12213/commits/5f5406ba04b75122205763402133c569f2324cb0)
    * https://github.com/symfony/symfony/issues/38627
        * [mocking event classes is not allowed anymore](https://github.com/mautic/mautic/pull/12213/commits/e9006ae93853b3910c2d65576b8a4809d482098a) 
        *  [workaround to test final ControllerEvent](https://github.com/mautic/mautic/pull/12213/commits/81d296a02e9fcb08f63e2803f20ea27f4df0fb8e)
    * [use InputBag for POST requests too, added missing scalar type hints](https://github.com/symfony/http-foundation/commit/7f7f5e3453c99d3e1f9c6c67edadae8015ed190e)
        * [workaround for inputbag not returning arrays](https://github.com/mautic/mautic/pull/12213/commits/b6a998bdf340d9caa32bac60e026e87c5bdd2e84)
        * [use InputBag instead of ParameterBag for request params](https://github.com/mautic/mautic/pull/12213/commits/3816a0dd3b65d88f0913fd4e8d0962e82257724a)
    *   [It is possible to pass request parameters of any type to the Symfony web crawler](https://github.com/symfony/symfony/issues/37740)
        * [fix tests related to the Symfony browser-kit changes](https://github.com/mautic/mautic/pull/12213/commits/4516979746b4043ba4c9c07e2d59eb38aeb04ee4)
        * [ensure that boolean parameter in tests are compatible with Symfony br…](https://github.com/mautic/mautic/pull/12213/commits/4dc13a34f95a60948e6c3ba9a4481b4d86fc0ed4)
    * method parameter of return type changes
        * [mock extra methods to ensure return value is correct](https://github.com/mautic/mautic/pull/12213/commits/c9f6099f9842ad68479ca2d50b5c59841fa7bdf7)
        * [update parameter types based on the extended method](https://github.com/mautic/mautic/pull/12213/commits/cc63c2635e631f4bb0d791983055fa0aaa0d2fae)
        * [update parameter and return types based on the extended method](https://github.com/mautic/mautic/pull/12213/commits/8e78ef7e88ef0f6a884669ad04fc1ae084a52260)
        * [adapt code to underlaying Symfony code](https://github.com/mautic/mautic/pull/12213/commits/06f9b3f0d7c4245a6a27549bafad0c62daac4753)

#### possible issues
* making the [`CoreSubscriberTes::testOnKernelController()`](https://github.com/mautic/mautic/pull/12213/commits/81d296a02e9fcb08f63e2803f20ea27f4df0fb8e) test work
    * Due to the `final` definition of `ControllerEvent` class, we can't mock it anymore.
    * Due to the way Mautic is built, there is no clear way to make this test do the same checks as now
    * possible solutions that didn't work
        * https://github.com/dg/bypass-finals doesn't solve the issue
        * creating a fake `ControllerEvent` class for test purposes has too much impact on the actual application code
    * the current implemented approach is to change as limited as possible the application code, but IMO this is a bad practice

###### Baseline of the tests errors and failures:

```
Tests: 3318, Assertions: 9600, Errors: 36, Failures: 11, Skipped: 5.
```

PHPSTAN:
```
Found 228 errors 
```

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. install Mautic via the UI
3. Install Mautic CLI
4. review the code changes to check if they are consistent
